Add syntax highlighting and LSP for Dockerfiles

everettraven opened the pull request:
Looks like it could resolve zed-industries/extensions#504 ? I haven't messed with any docker-compose stuff but seeing as that is YAML based and there is already YAML support this should address the Dockerfile syntax highlighting and LSP support.

I'm also a Rust newbie so if there is anything that looks off please do let me know!

A screenshot showing syntax highlighting when running via `cargo run`:
<img width="1226" alt="Screenshot 2024-01-27 at 8 28 25 PM" src="https://github.com/zed-industries/zed/assets/20310709/ebfd4c24-001f-4a8f-b085-e01f814cd1d9">


Release Notes:

- Added syntax highlighting and LSP for Dockerfiles

everettraven commented:
@cla-bot check

maxdeviant commented on crates/zed/src/languages/dockerfile.rs:
```diff
@@ -0,0 +1,124 @@
+    container_dir: PathBuf,
+    node: &dyn NodeRuntime,
+) -> Option<LanguageServerBinary> {
+    (|| async move {
```
We have an `async_maybe!` macro in our `util` crate that can make this a bit nicer than having the IIFE.

Here's an example of it used in another LSP adapter: https://github.com/zed-industries/zed/blob/5b9cc2619491c33b133254770ec2494010c2d016/crates/zed/src/languages/gleam.rs#L104

everettraven replied:
Oh nice, thanks for letting me know :) - done in the latest force push: https://github.com/zed-industries/zed/pull/6905/commits/e4e2b22da4754894adb384cfa4d7c7b52f6a28f2

maxdeviant commented on crates/zed/src/languages/dockerfile.rs:
```diff
@@ -0,0 +1,124 @@
+#[async_trait]
+impl LspAdapter for DockerfileLspAdapter {
+    fn name(&self) -> LanguageServerName {
+        LanguageServerName("dockerfile".into())
```
For the full language server name I think we'll want to use the name of the "binary":

```suggestion
        LanguageServerName("docker-langserver".into())
```

d1y commented:
Awesome, but I noticed that the syntax after PREFIX(run. etc) doesn't seem to be highlighted, I'm very impressed by the vscode plug-in, can you help me take a look: https://marketplace.visualstudio.com/items?itemName=jeff-hykin.better-dockerfile-syntax
<img width="940" alt="image" src="https://github.com/zed-industries/zed/assets/45585937/6aaf80c8-29eb-45f5-abd7-811a4ed52860">

SomeoneToIgnore commented:
This looks close to be done, so it would be great to add a new docs entry to the https://github.com/zed-industries/zed/tree/main/docs/src/languages list before merging.

everettraven committed:
Add syntax highlighting and LSP for Dockerfiles

Signed-off-by: everettraven <everettraven@gmail.com>

everettraven commented:
> Awesome, but I noticed that the syntax after PREFIX(run. etc) doesn't seem to be highlighted, I'm very impressed by the vscode plug-in, can you help me take a look: https://marketplace.visualstudio.com/items?itemName=jeff-hykin.better-dockerfile-syntax <img alt="image" width="940" src="https://private-user-images.githubusercontent.com/45585937/300230083-6aaf80c8-29eb-45f5-abd7-811a4ed52860.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY1MzkxNjAsIm5iZiI6MTcwNjUzODg2MCwicGF0aCI6Ii80NTU4NTkzNy8zMDAyMzAwODMtNmFhZjgwYzgtMjllYi00NWY1LWFiZDctODExYTRlZDUyODYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTI5VDE0MzQyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMzMTM2MzViMzM1Y2U2Y2ZmODYwMWYyM2ViNDI5YWM1NDlkMWUwM2MwNjVhZDRjMzMwNmMyM2I0ODgwYTIxYTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OO547GQDDU3TjM07VdebXax2gC3hBNuai77fYIOC_JE">

@d1y I updated the tree-sitter highlighting to get as close to this as I could. The new highlighting looks like this:
<img width="1229" alt="Screenshot 2024-01-29 at 9 32 04 AM" src="https://github.com/zed-industries/zed/assets/20310709/ff4bfef9-3f46-476f-b321-48cab18ba35e">

There is variable expansion highlighting where possible but there are some limitations as the tree-sitter-dockerfile crate doesn't currently have support for expansion in certain instructions (one is the `RUN` instruction) and no shell command syntax highlighting (it was either the whole shell command is highlighted or none, although this limitation could be due to me being more novice in messing with this tooling). That being said, I think this PR is "good enough" as is regarding the syntax highlighting and it could be incrementally improved in the future.

everettraven committed:
Add dockerfile to language docs

Signed-off-by: everettraven <everettraven@gmail.com>

everettraven commented:
> This looks close to be done, so it would be great to add a new docs entry to the https://github.com/zed-industries/zed/tree/main/docs/src/languages list before merging.

@SomeoneToIgnore Thanks for pointing that out! Rebased and added!

everettraven commented:
@maxdeviant @SomeoneToIgnore Anything else needed to get this merged?

SomeoneToIgnore commented on crates/zed/src/languages/dockerfile/config.toml:
```diff
@@ -0,0 +1,8 @@
+path_suffixes = ["Dockerfile"]
+line_comments = ["# "]
+brackets = [
+    { start = "{", end = "}", close = true, newline = true },
```
If we want syntax highlights for those, we need `brackets.scm`

SomeoneToIgnore reviewed (approved):
I looked at https://github.com/neondatabase/neon/blob/main/Dockerfile with this and looks solid, thank you for bringing this up.

One minor thing I see is this
![image](https://github.com/zed-industries/zed/assets/2690773/1ef6a439-eaf8-48c9-984e-899dcbbe9cf5)

I would expect ARG to be highlighted and both ARG and ENV values to be not highlighted.

I like what Intellij-based IDE do with bash code blocks and `--from` detection (e.g. recognizing `pg-build` and allowing to navigate to it)
![image](https://github.com/zed-industries/zed/assets/2690773/05c32738-83c0-4ef9-9864-7c40298dab62)

but that's advanced stuff somebody can work later (presumably, LSP improvements are needed for the navigation part?)

d1y commented:
Do you have any further questions about this PR? Can we merge? I really need this

SomeoneToIgnore commented:
> Can we merge? I

Would be nice, but GH does not allow this for the PRs with conflicts. Neither can I push into the branch, so one small step has to be done by somebody.

https://github.com/zed-industries/zed/pull/6905#issuecomment-1921258868

<img width="876" alt="image" src="https://github.com/zed-industries/zed/assets/2690773/5a3ec35b-432a-4779-8af8-1bc079e6228a">

SomeoneToIgnore commented:
https://github.com/zed-industries/zed/pull/7977 made that happen thanks to @d1y , hence closing this stale one.

everettraven commented:
@d1y Thanks for taking this over! I had some stuff come up and wasn't able to circle back around to address the comments/conflicts :(. Awesome to see this work get in!
